src/core/sandbox.c

```diff
@@ -0,0 +1,801 @@
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(ioctl), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* kill */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), SCMP_SYS(kill), 1,
+		SCMP_A1(SCMP_CMP_EQ, SIGTSTP));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), SCMP_SYS(kill), 1,
+		SCMP_A1(SCMP_CMP_EQ, SIGKILL));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(kill), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* prctl */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(prctl), 4,
+		SCMP_A0(SCMP_CMP_EQ, PR_SET_NAME),
+		SCMP_A2(SCMP_CMP_EQ, 0),
+		SCMP_A3(SCMP_CMP_EQ, 0),
+		SCMP_A4(SCMP_CMP_EQ, 0));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(prctl), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* fcntl */
+	const int fcntl_1arg_array[3] = {
+		F_GETFL,
+		F_GETFD,
+		F_SETLK
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(fcntl), 1,
+			SCMP_A1(SCMP_CMP_EQ, fcntl_1arg_array[i]));
+		if (rc < 0)
+			goto fail;
+	}
+
+	struct fcntl_2args_struct_t {
+		int cmd;
+		int arg;
+	};
+	const struct fcntl_2args_struct_t fcntl_2args_struct[3] = {
+		{ F_SETFD, FD_CLOEXEC          },
+		{ F_SETFL, O_RDONLY            },
+		{ F_SETFL, O_RDONLY|O_NONBLOCK }
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(fcntl), 2,
+			SCMP_A1(SCMP_CMP_EQ, fcntl_2args_struct[i].cmd),
+			SCMP_A2(SCMP_CMP_EQ, fcntl_2args_struct[i].arg));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(fcntl), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* mprotect */
+	const int mprotect_prot_array[3] = {
+		PROT_NONE,
+		PROT_READ,
+		PROT_READ|PROT_WRITE
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mprotect), 1,
+			SCMP_A2(SCMP_CMP_EQ, mprotect_prot_array[i]));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(mprotect), 60);
+	if (rc < 0)
+		goto fail;
+
+	/* mmap */
+	struct mmap_struct_t {
+		int prot;
+		int flags;
+	};
+	const struct mmap_struct_t mmap_anon_struct[4] = {
+		{ PROT_NONE,            MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS               },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK     },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS     }
+	};
+	for (i = 0; i < 4; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mmap), 4,
+			SCMP_A2(SCMP_CMP_EQ, mmap_anon_struct[i].prot),
+			SCMP_A3(SCMP_CMP_EQ, mmap_anon_struct[i].flags),
+			SCMP_A4(SCMP_CMP_EQ, 0xffffffff),
+			SCMP_A5(SCMP_CMP_EQ, 0));
+		if (rc < 0)
+			goto fail;
+	}
+
+	const struct mmap_struct_t mmap_noanon_struct[3] = {
+		{ PROT_READ,           MAP_PRIVATE               },
+		{ PROT_READ,           MAP_SHARED                },
+		{ PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE }
```

Comment:
Unfortunately, libseccomp does not allow for gradual tightening of the policy. As soon as the policy is enabled with seccomp_load(), it calls PR_SET_NO_NEW_PRIVS, although you can override that if you have CAP_SYS_ADMIN. This means that once a policy is enforced, it cannot be modified at all, including modified to be made more tight.

Waiting for library initialization would not be bad, although it does mean that seccomp would not work with plugins unless they are loaded right away. Perhaps waiting to for irssi to connect to the network before enabling seccomp? I can't think of all that many threat models where it would be important to protect irssi before it connects to the network.